crates/iceberg/src/io/object_cache.rs

```diff
@@ -0,0 +1,161 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use crate::io::FileIO;
+use crate::spec::{
+    FormatVersion, Manifest, ManifestFile, ManifestList, SchemaId, SnapshotRef, TableMetadataRef,
+};
+use crate::{Error, ErrorKind, Result};
+
+const DEFAULT_CACHE_SIZE_BYTES: u64 = 2 ^ 15; // 32MB
+
+#[derive(Clone, Debug)]
+pub(crate) enum CachedItem {
+    ManifestList(Arc<ManifestList>),
+    Manifest(Arc<Manifest>),
+}
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+pub(crate) enum CachedObjectKey {
+    ManifestList((String, FormatVersion, SchemaId)),
+    Manifest(String),
+}
+
+/// Caches metadata objects deserialized from immutable files
+#[derive(Clone, Debug)]
+pub struct ObjectCache {
+    cache: moka::future::Cache<CachedObjectKey, CachedItem>,
+    file_io: FileIO,
+    cache_disabled: bool,
+}
+
+impl ObjectCache {
+    /// Creates a new [`ObjectCache`]
+    /// with the default cache size
+    pub(crate) fn new(file_io: FileIO) -> Self {
+        Self::new_with_cache_size(file_io, DEFAULT_CACHE_SIZE_BYTES)
+    }
+
+    /// Creates a new [`ObjectCache`]
+    /// with a specific cache size
+    pub(crate) fn new_with_cache_size(file_io: FileIO, cache_size_bytes: u64) -> Self {
```

Comment:
We often use `capacity` for this. What do you think? For example, `ObjectCache::with_capacity`.

`ObjectCache::new_with_cache_size` doesn't look appealing to me.

Reply:
I like that! Done.